File: 조준장/level2/양궁 대회.py
def info_calculate(info, target):
    point = 0
    for i in range(len(info)):
        if info[i] == 0 and target[i] == 0: 
            continue 
        if info[i] < target[i]:
            point += (10 - i) 
        else:
            point -= (10 - i) 
    return point


def backtracking(info, result, start, n):
    global answer
    if n == 0 or start == 11:
        if n :
            result[-1] = n

        if info_calculate(info, answer) < info_calculate(info, result):
            answer = result[::]
        elif info_calculate(info, answer) == info_calculate(info, result):
            for i in range(11):
                if answer[10-i] < result[10-i]:
                    answer = result[::]
                    break
                elif answer[10-i] > result[10-i]:
                    break
        return
    
    for i in range(start, 11):
        if info[i] < n:
            result[i] = info[i] + 1
            backtracking(info, result, i+1, n-(info[i]+1))
            result[i] = 0
        
        elif i == 10:
            backtracking(info, result, i+1, n)
            result[-1] = 0

answer = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
def solution(n, info):
    backtracking(info, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 0, n)
    if info_calculate(info, answer) <= 0:
        return [-1]
    return answer

File: 이훈일/Programmers/Level2/양궁대회.java
class Solution {
    public int[] solution(int n, int[] info) {
        int dif = 0;
        int[] cur = new int[11];
        int[] result = new int[11];
        comb(n, 0, dif, cur, result, info);
        return result;
    }
    
    public static void comb(int n, int start, int dif, int[] cur, int[] result, int[] info) {
        
        if(start == 10) {
            return;
        }
        
        for(int i = start; i < cur.length; i++) {
            if(n >= info[i] + 1) {
                cur[i] = info[i]+ 1;
                n -= (info[i] + 1);
            }else {
                cur[10] = n;
                compare(dif, cur, result, info);
            }
            comb(n, start, dif, cur, result, info);
            n += cur[i];
            cur[i] = 0;
        }
    }
    
    public static void compare(int dif, int[] cur, int[] result, int[] info) {
        int curDif = 0;
        for(int i = 0; i < 11; i++) {
            if(info[i] < result[i]) {
                curDif += 10-i;                
            }else if(info[i] == result[i]){
                if(info[i] != 0) {
                    curDif -= 10-i;
                }
            }else {
                curDif -= 10-i;
            }
        }
        
        if(curDif > dif) {
            dif = curDif;
        }else if(curDif == dif) {
            for(int i = 10; i >= 0; i--) {
                if(cur[i] > result[i]) {
                    result = cur;
                }
            }
        }
    }
}
